matRad_SFUDoptimization.m

```diff
@@ -0,0 +1,136 @@
+            sb_cst{i,6}(j).dose = fx_dose * pln.numOfFractions;
+            
+        % physical dose splitting
+        else
+            sb_cst{i,6}(j).dose = sb_cst{i,6}(j).dose/pln.numOfBeams;
+        end
+    end
+end
+
+if ~isempty(dij)
+    % calculate dij with total dij being present
+        
+    % initialise total weight vector
+    wTot = zeros(dij.totalNumOfBixels,1);
+
+    for i = 1:pln.numOfBeams
+        % columns in total dij for single beam
+        sb_col = find(dij.beamNum == i);
+        % construct dij for single beam
+        sb_dij.numOfBeams = 1;
+        sb_dij.numOfVoxels = dij.numOfVoxels;
+        sb_dij.resolution = dij.resolution;
+        sb_dij.numOfRaysPerBeam = dij.numOfRaysPerBeam(i);
+        sb_dij.totalNumOfRays = sb_dij.numOfRaysPerBeam;
+        sb_dij.totalNumOfBixels = size(sb_col, 1);
+        sb_dij.dimensions = dij.dimensions;
+        sb_dij.numOfScenarios = dij.numOfScenarios;
+        sb_dij.bixelNum = dij.bixelNum(sb_col);
+        sb_dij.rayNum = dij.rayNum(sb_col);
+        sb_dij.beamNum = dij.beamNum(sb_col);
+        sb_dij.physicalDose{1} = dij.physicalDose{1}(:, sb_col);
+        if isfield(dij, 'RBE')
+            sb_dij.RBE = dij.RBE;
+        end
+        if isfield(dij, 'mLETDose')
+            sb_dij.mLETDose = dij.mLETDose(:, sb_col);
+        end
+        if isfield(dij,'mAlphaDose') && isfield(dij,'mSqrtBetaDose')
+            sb_dij.mAlphaDose{1} = dij.mAlphaDose{1}(:, sb_col);
+            sb_dij.mSqrtBetaDose{1} = dij.mSqrtBetaDose{1}(:, sb_col);
+        end
+
+        % adjust pln to one beam only
+        sb_pln = pln;
+        sb_pln.gantryAngles = pln.gantryAngles(i);
+        sb_pln.couchAngles = pln.couchAngles(i);
+
+        % optimize single beam
+        sb_resultGUI = matRad_fluenceOptimization(sb_dij,sb_cst,sb_pln);    
+
+        % merge single beam weights into total weight vector
+        wTot(sb_col) = sb_resultGUI.w;
+    end
+
+    % calculate dose
+    resultGUI = matRad_calcCubes(wTot,dij,cst,1);
+
+else
+    % calculate SFUD without total dij
+    
+    % initialise total weight vector
+    wTot = [];
+
+    for i = 1:pln.numOfBeams
+        fprintf(['optimizing beam ' num2str(i) '...\n']);
+        % single beam stf
+        sb_stf = stf(i);
+
+        % adjust pln to one beam only
+        sb_pln = pln;
+        sb_pln.isoCenter = pln.isoCenter(i,:);
+        sb_pln.numOfBeams = 1;
+        sb_pln.gantryAngles = pln.gantryAngles(i);
+        sb_pln.couchAngles = pln.couchAngles(i);
+
+        % calculate single beam dij
+        sb_dij = matRad_calcParticleDose(ct,sb_stf,sb_pln,sb_cst,false);
+
+        % optimize single beam
+        sb_resultGUI = matRad_fluenceOptimization(sb_dij,sb_cst,sb_pln);    
+
+        % merge single beam weights into total weight vector
+        wTot = cat(1,wTot,sb_resultGUI.w);
+
+    end
+
+    fprintf('Calculate total dose...\n');
+    % calculate dose
+    resultGUI = matRad_calcDoseDirect(ct,stf,pln,cst,wTot);
```

Comment:
But wouldn't we need the same amount of memory the total Dij matrix would take up as well?
This way one does not only save memory because no parts of the total dij are copied, but also because the maximum size of the sb_dij is way smaller than the total dij...
I think if you chose to do the optimisation without wanting to compute the total dij, you already went for the trade-off in favour of less memory usage and more computation time, so why not stick to that?